### PR TITLE
Add support for constructor handling in config beans

### PR DESCRIPTION
**Pull request by lodik**
Add support for beans, having not default constructor with annotation java.beans.ConstructorProperties. 
This allowing to write bean class using lombok annotation `@Value` instead of `@Data`

### CODE REVIEW

**Comment by radist-nt** [comment]
Great PR, with this change configuration beans can be finally made immutable!

The next change could be usage of constructor parameter names instead of specifying them in `@ConstructorProperties` annotation for code compiled with parameter names (with `-parameters` compiler argument). But the annotation is still required to specify the constructor used in `ConfigBeanImpl`. That's why, I suggest to rename the annotation to `@ConfigConstructor`, because it's primary purpose is to specify the constructor used in `ConfigBeanImpl`.

**Comment by lodik** [comment]
@havocp Can you take a look?

File: README.md [hunk 1]
```diff
@@ -482,9 +482,11 @@ and
 
 ### ConfigBeanFactory
 
-As of version 1.3.0, if you have a Java object that follows
-JavaBean conventions (zero-args constructor, getters and setters),
-you can automatically initialize it from a `Config`.
+You can initialize some class from a `Config`, if a class contains 
+a no-arg constructor, or a public constructor annotated with `java.beans.ConstructorProperties` 
+or a constructor which have a parameter names. The constructor parameter name will be used to
```
Comment (radist-nt) [inline README.md:487]: or class compiled with `-parameters` compiler argument.

**Comment by Gaibhne** [comment]
@havocp what can/needs to be done to get this PR merged ? At my place we try to aggressively use immutable objects to lower risk of bugs, so this would help us a lot

**Comment by Gaibhne** [comment]
@havocp @radist-nt  I hate to 'ping' you like this, but it is unclear why this (almost two year old!) user contribution has not been merged or rejected. Could you please clear up what is going on or what can be done ?

**Comment by havocp** [comment]
There are tons of open issues and PRs, in general I don't have time to work on them for now. Sometimes I will approve something or comment on something that's really easy but this one would take me some time to review in a meaningful way. It's possible someone at Lightbend has some time to look at it.

(larger discussion, I think the library would benefit from a couple months of focused work to catch up on all the issues and PRs but I'm not sure how to make the stars align so someone with the right knowledge has that time available.)

**Comment by Gaibhne** [comment]
Thanks for the response, while disappointing, it is absolutely understandable for an open source project to be lacking in resources.

**Comment by ennru** [comment]
We do not intend to extend the functionality of "Typesafe Config" further.
See https://github.com/lightbend/config#maintained-by

**Pull request closed without merging**